webui/react/src/pages/Cluster.tsx

```diff
@@ -0,0 +1,55 @@
+import React, { useCallback } from 'react';
+
+import Grid from 'components/Grid';
+import Page from 'components/Page';
+import ResourceChart from 'components/ResourceChart';
+import Spinner from 'components/Spinner';
+import AgentsCtx from 'contexts/Agents';
+import emptyMessage from 'styles/emptyMessage.module.scss';
+import { Resource } from 'types';
+import { categorize } from 'utils/data';
+
+import css from './Cluster.module.scss';
+
+const Cluster: React.FC = () => {
+  const agents = AgentsCtx.useStateContext();
+
+  const getResources = useCallback(() => {
+    if (!agents.data) return {};
+    const resourceList = agents.data?.map(a => a.resources).reduce((a, c) => {
+      a.push(...c); return a;
+    });
+    return categorize(resourceList, (res: Resource) => res.type);
+  }, [ agents ]);
+
+  const resourceDict = getResources();
+
+  const availableResTypes = Object.keys(resourceDict);
+
+  let unhappyView: React.ReactNode;
```

Comment:
suggestion: default `unhappyView` to `<Spinner />` during declaration and initialization

Reply:
thought: also in the case where you don't want to render anything (unlike the case above), I think it is recommended that you pass `null` instead of an `undefined`, but I think either works. So explicitly, if we don't want to render a spinner by default above, we would make the declaration like this instead:
```
let unhappyView: React.ReactNode = null;
```